### ISSUE TITLE
Potential bug in Operator State management

### ISSUE THREAD

**Issue by kumare3** [GoogleCloudPlatform/spark-on-k8s-operator#199]
The operator seems to monitor events (informer) for the CRD and the executor/driver pods independently. When a pod is updated the state is preserved in etcd using optimistic concurrency. The problem is, we do not do a CAS style operation, which ensures that a previously successful/completion job does not go back to running.
Let us take an example,
 1. Job is submitted, state is submitted
 2. driver goes to running, but fails write and waits before a retry
 3. driver completes and status update to complete is successful
 4. now the previously pending write is applied, which causes the state to be reverted to submitted/running
Basically the problem is out of order updates being applied to a consistent data store
The manifestation is that the app state get stuck in a non terminal state and affects the observer.

We have a simple hack to resolve this issue for now
https://github.com/lyft/spark-on-k8s-operator/blob/spark-2.2-support/pkg/controller/sparkapplication/controller.go#L521-L524

A problem with the hack is we can lose the executor states (as we just ignore non terminal states)
We will work on cleaning up this and upstreaming the fix if you think this is a problem

**Comment by kumare3** [comment]
On a side note, we could simplify the state machine management, by making sure all the updates to the state happen in the same thread (goroutine). For example, any pod update (for spark) just triggers the top level CRD's onUpdate and then we use the listers to determine and take the right action. @liyinan926 we would love to discuss this and help you refactor the code. For now we will implement the bug fix and share that with you.

**Comment by liyinan926** [comment]
@kumare3 It's a good point that state updates should be handled by a single thread in a defined order, e.g., FIFO. We could trigger the top-level `OnUpdate` whenever there's an status update of the driver or executor pods. Specifically, whenever the controller receives a driver or executor pod status update through the `appStateReportingChan` channel, the key of the corresponding `SparkApplication` object gets enqueued. By doing this, however, instead of updating the status of a `SparkApplication` based on the received status updates of the driver or executor pods, within `OnUpdate` the operator needs to list the driver and executor pods and update the application status based on the result of the listing. This is much closer to what other custom controllers/operators do to achieve to so-called "level-based triggering".

**Comment by kumare3** [comment]
@liyinan926 yes that is what I was suggesting. We have this implemented in a few Operators internally (which we plan to open source) We would love to contribute this if you think this could be a great addition. I feel this would simplify the operator tremendously and will help us add new features much faster? Lets discuss this during our call tomorrow.

**Comment by liyinan926** [comment]
@kumare3 Yes, I think this would be some great addition!
